### PR TITLE
sdk/metric/metricdata: Add MarshalJSON for Extrema

### LINKED ISSUE #4826: I cannot Get Max/Min Value In Histogram type
### Description

I noticed that In DefaultAggregationSelector, the max/min value are recorded by default.  But when I tried to use it with a simple stdout exporter, it comes that min&max are always empty. How Should I use it? is there any example?

image

image

```go 
func DefaultAggregationSelector(ik InstrumentKind) Aggregation {
	switch ik {
	case InstrumentKindCounter, InstrumentKindUpDownCounter, InstrumentKindObservableCounter, InstrumentKindObservableUpDownCounter:
		return AggregationSum{}
	case InstrumentKindObservableGauge:
		return AggregationLastValue{}
	case InstrumentKindHistogram:
		return AggregationExplicitBucketHistogram{
			Boundaries: []float64{0, 5, 10, 25, 50, 75, 100, 250, 500, 750, 1000, 2500, 5000, 7500, 10000},
			NoMinMax:   false,
		}
	}
	panic("unknown instrument kind")
}
```

### Environment

- OS: macOS Big Sur (11.x)
- Architecture: arm64
- Go Version: 1.20
- opentelemetry-go version: 1.21.0

### Steps To Reproduce

```go
package main

import (
	"context"
	"go.opentelemetry.io/otel"
	"go.opentelemetry.io/otel/exporters/stdout/stdoutmetric"
	otelmetric "go.opentelemetry.io/otel/metric"
	"go.opentelemetry.io/otel/sdk/metric"
	"go.opentelemetry.io/otel/sdk/resource"
	semconv "go.opentelemetry.io/otel/semconv/v1.21.0"
	"log"
	"math/rand"
	"net/http"
	"time"
)

func MockOpenTelemetry() {
	// Create resource.
	res, err := newResource()
	if err != nil {
		panic(err)
	}
	meterProvider, err := newMeterProvider(res)
	if err != nil {
		panic(err)
	}

	defer func() {
		if err := meterProvider.Shutdown(context.Background()); err != nil {
			log.Println(err)
		}
	}()

	otel.SetMeterProvider(meterProvider)

	var meter = otel.Meter("my-service-meter")
	histogram, err := meter.Int64Histogram(
		"task.duration",
		otelmetric.WithDescription("The duration of task execution."),
		otelmetric.WithUnit("ms"),
	)
	if err != nil {
		panic(err)
	}
	http.HandleFunc("/", func(w http.ResponseWriter, r *http.Request) {
		start := time.Now()

		// do some work in an API call
		time.Sleep(100*time.Millisecond + time.Duration(rand.Intn(901))*time.Millisecond)

		duration := time.Since(start)
		histogram.Record(r.Context(), duration.Milliseconds())
	})
	log.Fatal(http.ListenAndServe(":8080", nil))
}

func newResource() (*resource.Resource, error) {
	return resource.Merge(resource.Default(),
		resource.NewWithAttributes(semconv.SchemaURL,
			semconv.ServiceName("my-service"),
			semconv.ServiceVersion("0.1.0"),
		))
}

func newMeterProvider(res *resource.Resource) (*metric.MeterProvider, error) {
	metricExporter, err := stdoutmetric.New(stdoutmetric.WithTemporalitySelector(deltaSelector), stdoutmetric.WithAggregationSelector(metric.DefaultAggregationSelector))
	if err != nil {
		return nil, err
	}

	meterProvider := metric.NewMeterProvider(
		metric.WithResource(res),
		metric.WithReader(metric.NewPeriodicReader(metricExporter,
			// Default is 1m. Set to 3s for demonstrative purposes.
			metric.WithInterval(1*time.Minute))),
	)
	return meterProvider, nil
}

```

{"Resource":[{"Key":"service.name","Value":{"Type":"STRING","Value":"my-service"}},{"Key":"service.version","Value":{"Type":"STRING","Value":"0.1.0"}},{"Key":"telemetry.sdk.language","Value":{"Type":"STRING","Value":"go"}},{"Key":"telemetry.sdk.name","Value":{"Type":"STRING","Value":"opentelemetry"}},{"Key":"telemetry.sdk.version","Value":{"Type":"STRING","Value":"1.21.0"}}],"ScopeMetrics":[{"Scope":{"Name":"my-service-meter","Version":"","SchemaURL":""},"Metrics":[{"Name":"task.duration","Description":"The duration of task execution.","Unit":"ms","Data":{"DataPoints":[{"Attributes":[],"StartTime":"2024-01-14T13:08:48.504319+08:00","Time":"2024-01-14T13:09:48.50817+08:00","Count":12,"Bounds":[0,5,10,25,50,75,100,250,500,750,1000,2500,5000,7500,10000],"BucketCounts":[0,0,0,0,0,0,0,3,1,3,5,0,0,0,0,0],"Min":{},"Max":{},"Sum":6930}],"Temporality":"DeltaTemporality"}}]}]}
{"Resource":[{"Key":"service.name","Value":{"Type":"STRING","Value":"my-service"}},{"Key":"service.version","Value":{"Type":"STRING","Value":"0.1.0"}},{"Key":"telemetry.sdk.language","Value":{"Type":"STRING","Value":"go"}},{"Key":"telemetry.sdk.name","Value":{"Type":"STRING","Value":"opentelemetry"}},{"Key":"telemetry.sdk.version","Value":{"Type":"STRING","Value":"1.21.0"}}],"ScopeMetrics":[{"Scope":{"Name":"my-service-meter","Version":"","SchemaURL":""},"Metrics":[{"Name":"task.duration","Description":"The duration of task execution.","Unit":"ms","Data":{"DataPoints":[{"Attributes":[],"StartTime":"2024-01-14T13:09:48.50817+08:00","Time":"2024-01-14T13:10:48.506979+08:00","Count":4,"Bounds":[0,5,10,25,50,75,100,250,500,750,1000,2500,5000,7500,10000],"BucketCounts":[0,0,0,0,0,0,0,3,0,0,1,0,0,0,0,0],"Min":{},"Max":{},"Sum":1452}],"Temporality":"DeltaTemporality"}}]}]}

### Expected behavior

I expecte to get max/min value in histogram with delta temporality

### PR DESCRIPTION
**Pull request by m-posluszny**
Fixes #4826 

### Issue

Json representation was missing Max and Min values, due to private fields of `Extrema` type.

> [{"Attributes":[],"StartTime":"2024-01-14T13:09:48.50817+08:00","Time":"2024-01-14T13:10:48.506979+08:00","Count":4,"Bounds":[0,5,10,25,50,75,100,250,500,750,1000,2500,5000,7500,10000],"BucketCounts":[0,0,0,0,0,0,0,3,0,0,1,0,0,0,0,0],"Min":{},"Max":{},"Sum":1452}],"Temporality":"DeltaTemporality"}}]

### Solution

Implemented MarshalJSON method, that converts field `value` of `Extrema` to BJSON number type. In case of invalid number, 0 will be returned.

### Does it work?

I've tested it over example provided in mentioned issue. I've managed to get min and max values of time durations. Moreover I've added unit tests to cover json marshalling  for `Extrema` type.

### CODE REVIEW

File: Makefile [hunk 1]
```diff
@@ -192,7 +192,7 @@ test-coverage: | $(GOCOVMERGE)
 	done; \
 	$(GOCOVMERGE) $$(find . -name coverage.out) > coverage.txt
 
-# Adding a directory will include all benchmarks in that direcotry if a filter is not specified.
+# Adding a directory will include all benchmarks in that directory if a filter is not specified.
```
Comment (MrAlias) [inline Makefile:195]: Please submit this as it's own PR. It is unrelated to these changes.

File: sdk/metric/metricdata/data.go [hunk 2]
```diff
@@ -211,6 +213,14 @@ type Extrema[N int64 | float64] struct {
 	valid bool
 }
 
+// MarshalJson converts the Extrema value to JSON number
+func (e *Extrema[N]) MarshalJSON() ([]byte, error) {
```
Comment (MrAlias) [inline sdk/metric/metricdata/data.go:217]: Implementing `MarshalText` will be more general and support more use cases. Can we implement that instead?

**Review by MrAlias — COMMENTED** [review]
Please sign the CLA so these changes can be considered for merge.

File: sdk/metric/metricdata/data.go (hunk 2 above)
Comment (m-posluszny) [reply sdk/metric/metricdata/data.go:217]: I've started with `MarshalText` but moved to `MarshalJSON` to keep API consistent. Rest of the numerical values are represented as json `number`, which cannot be done with `MarshalText` as json encoder will translate it as `string`. 

Can we agree on keeping `MarshalJSON` and implementing `MarshalText` next to it? 
As per [json docs](https://pkg.go.dev/encoding/json#Marshal), if `MarshalJSON` is implemented it will be preferred over `MarshalText` for json encoding. So we'll have no issues with JSON types and `Extrema` type will still be represented in encoders that utilize `MarshalText`

File: sdk/metric/metricdata/data.go [hunk 3]
```diff
@@ -211,6 +213,14 @@ type Extrema[N int64 | float64] struct {
 	valid bool
 }
 
+// MarshalJson converts the Extrema value to JSON number
```
Comment (pellared) [inline sdk/metric/metricdata/data.go:216]:
```suggestion
// MarshalJSON converts the Extrema value to JSON number.
```

File: sdk/metric/metricdata/data.go [hunk 4]
```diff
@@ -211,6 +213,14 @@ type Extrema[N int64 | float64] struct {
...
+func (e *Extrema[N]) MarshalJSON() ([]byte, error) {
+	if !e.valid {
+		return []byte("0"), nil
+	}
+	return []byte(json.Number(fmt.Sprint(e.value))), nil
```
Comment (pellared) [inline sdk/metric/metricdata/data.go:221]: Wy not simply?

```suggestion
	return json.Marshal(e.value)
```

Comment (m-posluszny) [reply sdk/metric/metricdata/data.go:221]: Well, not really. I've started with implementation of MarshalText, then refactored it to MarshalJSON so I wasn't thinking on replacing the whole statement. I'm waiting on resolution of [discussion](https://github.com/open-telemetry/opentelemetry-go/pull/4827#discussion_r1453182773) regarding MarshalText, so I'll hold back with simplifying that one till it's resolved.

Comment (pellared) [reply sdk/metric/metricdata/data.go:221]: It is resolved. You can address this and other comments.

**Comment by pellared** [comment]
Can you add a changelog entry?

File: CHANGELOG.md [hunk 5]
```diff
@@ -24,6 +24,7 @@ This project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.htm
 - Experimental cardinality limiting is added to the metric SDK.
   See [metric documentation](./sdk/metric/EXPERIMENTAL.md#cardinality-limit) for more information about this feature and how to enable it. (#4457)
 - Add `NewMemberRaw` and `NewKeyValuePropertyRaw` in `go.opentelemetry.io/otel/baggage`. (#4804)
+- Add `MarshalText` and `MarshalJSON` in `go.opentelemetry.io/otel/sdk/metric/metricdata`.  (#4826)
```
Comment (pellared) [inline CHANGELOG.md:27]: 1. This should be in `Unrealeased` section.
2. The `Extrema` type is not mentioned
3. I think we should also say that it fixes the missing Min and Max values for `go.opentelemetry.io/otel/exporters/stdout/stdoutmetric` (therefore I would rather put it in `Fixed` than `Added`).

File: CHANGELOG.md [hunk 6]
```diff
@@ -8,6 +8,10 @@ This project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.htm
 
 ## [Unreleased]
 
+### Fixed
+
+- Fixed missing Mix and Max values for `go.opentelemetry.io/otel/exporters/stdout/stdoutmetric` by introducing `MarshalText` and `MarshalJSON` for type `Extrema` in `go.opentelemetry.io/sdk/metric/metricdata` (#4826)
```
Comment (pellared) [inline CHANGELOG.md:13]:
```suggestion
- Fixed missing Mix and Max values for `go.opentelemetry.io/otel/exporters/stdout/stdoutmetric` by introducing `MarshalText` and `MarshalJSON` for type `Extrema` in `go.opentelemetry.io/sdk/metric/metricdata`. (#4826)
```

Comment (pellared) [inline CHANGELOG.md:13]:
```suggestion
- Fixed missing Mix and Max values for `go.opentelemetry.io/otel/exporters/stdout/stdoutmetric` by introducing `MarshalText` and `MarshalJSON` for type `Extrema` in `go.opentelemetry.io/sdk/metric/metricdata`. (#4826)
```

File: sdk/metric/metricdata/data.go [hunk 7]
```diff
@@ -211,6 +212,19 @@ type Extrema[N int64 | float64] struct {
 	valid bool
 }
 
+// MarshalText converts the Extrema value to text.
+func (e Extrema[N]) MarshalText() ([]byte, error) {
+	if !e.valid {
+		return []byte("0"), nil
```
Comment (pellared) [inline sdk/metric/metricdata/data.go:218]: @MrAlias is it a correct reprentation for an invalid extrema? Shouldn't we return nil, nil?

**Comment by pellared** [comment]
Unit tests in `stdoutmetric` need to be fixed/updated.

File: sdk/metric/metricdata/data.go (hunk 7 above)
Comment (m-posluszny) [reply sdk/metric/metricdata/data.go]: @MrAlias nil will output empty bytes, which will break MarshalJSON. Can we settle on `return json.Marshal(nil)` for `MarshalText`? 
Text representation will then be equal to `null`, not to just to no bytes.

**Comment by m-posluszny** [comment]
Sorry for the force push, had to reverse problematic rebase

File: exporters/stdout/stdoutmetric/example_test.go [hunk 8]
```diff
@@ -257,8 +257,8 @@ func Example() {
 	//                   6,
 	//                   0
 	//                 ],
-	//                 "Min": {},
-	//                 "Max": {},
+	//                 "Min": null,
+	//                 "Max": null,
```
Comment (pellared) [inline exporters/stdout/stdoutmetric/example_test.go:261]: nit: I think we can update the `mockData` to have meaningful example.

**Review by pellared — APPROVED** [review]

File: exporters/stdout/stdoutmetric/example_test.go (hunk 8 above)
Comment (m-posluszny) [reply exporters/stdout/stdoutmetric/example_test.go:320]: I've added separate entry called `requests.size` as I think we should have one example without specific values to check for proper `nil` marshaling.

File: sdk/metric/metricdata/metricdatatest/assertion_test.go [hunk 9]
```diff
@@ -1010,3 +1013,22 @@ func TestAssertAttributesFail(t *testing.T) {
...
+func AssertMarshal[N int64 | float64](t *testing.T, expected string, i *metricdata.Extrema[N]) {
+	t.Helper()
+
+	b, err := json.Marshal(i)
+	assert.NoError(t, err)
+	assert.Equal(t, expected, string(b))
+}
+
+func TestAssertMarshal(t *testing.T) {
+	AssertMarshal(t, "-1", &minFloat64A)
```
Comment (MrAlias) [inline sdk/metric/metricdata/metricdatatest/assertion_test.go:1026]: Missing test for empty `Extrema`. I.e.

```suggestion
	AssertMarshal(t, "", &metricdata.Extream{})
	AssertMarshal(t, "-1", &minFloat64A)
```

Comment (m-posluszny) [reply sdk/metric/metricdata/metricdatatest/assertion_test.go:1028]: empty `Extrema` test added

**Review by MrAlias — APPROVED** [review]

**Comment by MadVikingGod** [comment]
I'm on board with adding this.  Should we also include a disclaimer that this is the go internal format and will not be compatible with other languages SDK, and for that, they should use OTLP?

**Comment by MrAlias** [comment]
> I'm on board with adding this. Should we also include a disclaimer that this is the go internal format and will not be compatible with other languages SDK, and for that, they should use OTLP?

@MadVikingGod this is already documented in the stdout exporter: https://github.com/open-telemetry/opentelemetry-go/blob/e6e4c95d88ca1ce4937996d522130ff592c6fdbb/exporters/stdout/stdoutmetric/doc.go#L18-L22

Are you talking about something more?

**Review by MadVikingGod — APPROVED** [review]
